### PR TITLE
[GSB] Ensure that term rewriting rules produce correct anchors

### PR DESCRIPTION
**Pull request by DougGregor**
Rework the way that we form term rewriting rules in the `GenericSignatureBuilder`, such that we form rewrite rules only when merging two equivalence classes, in which case we form a rewrite rule from the anchor of one equivalence class to the anchor of the other equivalence class. This produces smaller sets of rewrite rules (by far) and ensures that we always canonicalize to the anchor.

While here, fix a problem where we failing to apply rewrite rules like tau_0_1 -> tau_0_0. With both changes in, introduce assertions to make sure we're getting a proper anchor.

Fixes rdar://problem/37469390.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/AST/GenericSignatureBuilder.cpp [hunk 1]
```diff
@@ -2157,19 +2149,42 @@ Type EquivalenceClass::getAnchor(
...
+
     // Record the cache miss and update the cache.
     ++NumArchetypeAnchorCacheMisses;
     archetypeAnchorCache.anchor = anchorType;
     archetypeAnchorCache.lastGeneration = builder.Impl->Generation;
-    return substAnchor();
+    updatedAnchor = true;
+
+#if NDEBUG
+    break;
```
Comment (airspeedswift) [inline lib/AST/GenericSignatureBuilder.cpp]: @DougGregor this introduced a warning in no-assert builds about the loop only going round once. If you have a preferred suggested way to fix that, I'll raise a PR to do it.

**Pull request merged**